### PR TITLE
Enhance Pulsar Connect api by attaching more information about the message

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation

Instead of just passing the message data, it might be useful to have other attributes of the message that is being read/written from/to external connectors. To that end, this pr enhances the Pulsar Connect interface by adding a Message class to encapsulate additional information

### Modifications

Describe the modifications you've done.

### Result

After your change, what will change.

### CODE REVIEW

**Comment by srkukarni** [comment]
@sijie @jerrypeng

File: pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/Message.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.connect.core;
+
+/**
+ * Pulsar Connect's Message interface. Message encapsulates the
+ * information about a message being read/written from/to a Source/Sink.
+ */
+public class Message<T> {
```
Comment (jerrypeng) [inline pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/Message.java]: Can we make this an interface and a corresponding implementation

Comment (srkukarni) [reply pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/Message.java]: Done

File: pulsar-connect/twitter/src/main/java/org/apache/pulsar/connect/twitter/TwitterFireHose.java [hunk 2]
```diff
@@ -164,4 +165,27 @@ private void stopThread() {
         }
     }
 
+    private class TwitterMessage implements Message<String> {
```
Comment (sijie) [inline pulsar-connect/twitter/src/main/java/org/apache/pulsar/connect/twitter/TwitterFireHose.java]: nit: make static

File: pulsar-connect/rabbitmq/src/main/java/org/apache/pulsar/connect/rabbitmq/RabbitMQSource.java [hunk 3]
```diff
@@ -74,16 +75,39 @@ public void close() throws Exception {
...
         }
 
         @Override
         public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body) throws IOException {
-            consumeFunction.apply(body);
+            consumeFunction.apply(new RabbitMQMessage(body));
+        }
+    }
+
+    private class RabbitMQMessage implements Message<byte[]> {
```
Comment (sijie) [inline pulsar-connect/rabbitmq/src/main/java/org/apache/pulsar/connect/rabbitmq/RabbitMQSource.java]: nit: make static

File: pulsar-connect/kafka/src/main/java/org/apache/pulsar/connect/kafka/KafkaSource.java [hunk 4]
```diff
@@ -120,7 +121,30 @@ public void start() {
     }
 
     @Override
-    public void setConsumer(java.util.function.Function<V, CompletableFuture<Void>> consumeFunction) {
+    public void setConsumer(java.util.function.Function<Message<V>, CompletableFuture<Void>> consumeFunction) {
         this.consumeFunction = consumeFunction;
     }
+
+    private class KafkaMesssage implements Message<V>  {
```
Comment (sijie) [inline pulsar-connect/kafka/src/main/java/org/apache/pulsar/connect/kafka/KafkaSource.java]: nit: make static

File: pulsar-connect/twitter/src/main/java/org/apache/pulsar/connect/twitter/TwitterFireHose.java (hunk 2 above)
Comment (srkukarni) [reply pulsar-connect/twitter/src/main/java/org/apache/pulsar/connect/twitter/TwitterFireHose.java]: done

File: pulsar-connect/rabbitmq/src/main/java/org/apache/pulsar/connect/rabbitmq/RabbitMQSource.java (hunk 3 above)
Comment (srkukarni) [reply pulsar-connect/rabbitmq/src/main/java/org/apache/pulsar/connect/rabbitmq/RabbitMQSource.java]: done

File: pulsar-connect/kafka/src/main/java/org/apache/pulsar/connect/kafka/KafkaSource.java (hunk 4 above)
Comment (srkukarni) [reply pulsar-connect/kafka/src/main/java/org/apache/pulsar/connect/kafka/KafkaSource.java]: done

File: pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/Message.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.connect.core;
+
+/**
+ * Pulsar Connect's Message interface. Message encapsulates the
+ * information about a message being read/written from/to a Source/Sink.
+ */
+public interface Message<T> {
```
Comment (jerrypeng) [inline pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/Message.java]: @srkukarni I was think the Message interface should look like the following:

public interface Message<T> {

    static  <V> Message<V> create(String partitionId, Long sequenceId, V data) {
        return new Message<V>() {
            @Override
            public String getPartitionId() {
                return partitionId;
            }

            @Override
            public Long getSequenceId() {
                return sequenceId;
            }

            @Override
            public V getData() {
                return data;
            }
        };
    }
    /**
     * Retrieves the partition information if any of the message
     * @return The partition id where the
     */
    String getPartitionId();

    /**
     * Retrieves the sequence id of the message
     * @return Sequence Id associated with the message
     */
    Long getSequenceId();

    /**
     * Retrieves the actual data of the message
     * @return The message data
     */
    T getData();
}

Comment (jerrypeng) [reply pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/Message.java]: Users can just use the static create method in the interface

**Comment by jerrypeng** [comment]
LGTM +1

**Comment by srkukarni** [comment]
retest this please

**Comment by srkukarni** [comment]
retest this please

**Pull request merged**